src/shred_collector/shred_receiver.zig

```diff
@@ -0,0 +1,202 @@
+const std = @import("std");
+const network = @import("zig-network");
+const sig = @import("../lib.zig");
+
+const bincode = sig.bincode;
+const layout = sig.shred_collector.shred_layout;
+
+const Allocator = std.mem.Allocator;
+const ArrayList = std.ArrayList;
+const Atomic = std.atomic.Value;
+const KeyPair = std.crypto.sign.Ed25519.KeyPair;
+const Socket = network.Socket;
+
+const Channel = sig.sync.Channel;
+const Logger = sig.trace.Logger;
+const Packet = sig.net.Packet;
+const Ping = sig.gossip.Ping;
+const Pong = sig.gossip.Pong;
+const RepairMessage = sig.shred_collector.RepairMessage;
+const Slot = sig.core.Slot;
+const SocketThread = sig.net.SocketThread;
+
+const endpointToString = sig.net.endpointToString;
+
+/// Analogous to [ShredFetchStage](https://github.com/anza-xyz/agave/blob/aa2f078836434965e1a5a03af7f95c6640fe6e1e/core/src/shred_fetch_stage.rs#L34)
+pub const ShredReceiver = struct {
+    allocator: Allocator,
+    keypair: *const KeyPair,
+    exit: *Atomic(bool),
+    logger: Logger,
+    repair_socket: Socket,
+    tvu_socket: Socket,
+    outgoing_shred_channel: *Channel(ArrayList(Packet)),
+    shred_version: *const Atomic(u16),
+
+    const Self = @This();
+
+    /// Run threads to listen/send over socket and handle all incoming packets.
+    /// Returns when exit is set to true.
+    pub fn run(self: *Self) !void {
+        defer self.logger.err("exiting shred receiver");
+        errdefer self.logger.err("error in shred receiver");
+
+        var sender = try SocketThread
+            .initSender(self.allocator, self.logger, self.repair_socket, self.exit);
+        defer sender.deinit();
+        var repair_receiver = try SocketThread
+            .initReceiver(self.allocator, self.logger, self.repair_socket, self.exit);
+        defer repair_receiver.deinit();
+
+        const num_tvu_receivers = 2;
+        var tvu_receivers: [num_tvu_receivers]*Channel(ArrayList(Packet)) = undefined;
+        for (0..num_tvu_receivers) |i| {
+            tvu_receivers[i] = (try SocketThread.initReceiver(
+                self.allocator,
+                self.logger,
+                self.tvu_socket,
+                self.exit,
+            )).channel;
+        }
+        defer for (tvu_receivers) |r| r.deinit();
+        const x = try std.Thread.spawn(
+            .{},
+            Self.runPacketHandler,
+            .{ self, tvu_receivers, sender.channel },
+        );
+        const y = try std.Thread.spawn(
+            .{},
+            Self.runPacketHandler,
+            .{ self, .{repair_receiver.channel}, sender.channel },
+        );
+        x.join();
+        y.join();
+    }
+
+    /// Keep looping over packet channel and process the incoming packets.
+    /// Returns when exit is set to true.
+    fn runPacketHandler(
+        self: *Self,
+        receivers: anytype,
```

Comment:
The channels are already pointers. Using a slice would mean I'm passing a pointer to a pointer. I wanted to avoid the extra layer of indirection. I don't know if if the indirection has any actual performance consequences, but I figure it definitely doesn't help, and it could potentially hurt.

To avoid indirection, the channel pointers can be passed by value as an array or tuple. This would either require either an comptime parameters for the function, or making this parameter anytype.

I see that this does make this harder to read, so I'm not opposed to using a slice instead.

Reply:
> (&1)[0..1]

I was not aware of this syntax until I saw it in Trevor's pr. In the past I used `&.{item}` for this purpose. Is there a difference in semantics between the two?

Reply:
yeah if theres no strong reason for, would pref slice to make it more readable

Reply:
0544591663d80bf8307ac38f51d500b2e13cb408